### DISCUSSION TITLE
[@types/d3-selection] sub-selection types are wrong

### DISCUSSION THREAD

**Discussion by itsmebasti** [DefinitelyTyped/DefinitelyTyped#65932] [Issues with a @types package]
I am working with d3 and typescript and I am confused about the type that is returned from various methods of a Selection.

This is the current [selection interface signature][1]:
```ts
/**
 * A D3 Selection of elements.
 *
 * The first generic "GElement" refers to the type of the selected element(s).
 * The second generic "Datum" refers to the type of the datum of a selected element(s).
 * The third generic "PElement" refers to the type of the parent element(s) in the D3 selection.
 * The fourth generic "PDatum" refers to the type of the datum of the parent element(s).
*/
export interface Selection<GElement extends BaseType, Datum, PElement extends BaseType, PDatum> {
```
now this interface has several functions defined that return sub-selections of this selection and most of them return the wrong type, only two are correct in my understanding. Let me show two examples

selectAll (correct)
```ts
selectAll<DescElement extends BaseType, OldDatum>(selector: string)
    : Selection<DescElement, OldDatum, GElement, Datum>;
```
select (incorrect)
```ts
select<DescElement extends BaseType>(selector: string)
    : Selection<DescElement, Datum, PElement, PDatum>;
```
So instead of passing its own type into the sub-select's type, the function seems to pass in its parent's type und datum.

This is a complete list of the affected functions (overloads are the same)

    select         //incorrect
    selectAll
    selectChild    //incorrect
    selectChildren
    append         //incorrect
    insert         //incorrect

I am confused if I didn't understand the concept since the d3 types are downloaded a million times weekly. So I am not sure how to proceed here, and if/where to fix this. Should I open a PR against DefinitelyTyped, or is someone else managing those types?

  [1]: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/d172d3e076f25c468c0177896d9369f23c45259c/types/d3-selection/index.d.ts

[a comment by an automated account is not shown]

**Comment by Methuselah96** [comment]
You could very well be correct. The best way to verify would be to create a runnable example where we could test it out. Feel free to submit a PR!
